salt/transport/ipc.py

```diff
@@ -0,0 +1,305 @@
+        '''
+        # Start up the ioloop
+        log.trace('IPCServer: binding to socket: {0}'.format(socket_path))
+        self.sock = tornado.netutil.bind_unix_socket(socket_path)
+
+        if self.io_loop is None:
+            self.io_loop = IOLoop.current()
+
+        tornado.netutil.add_accept_handler(
+            self.sock,
+            self.handle_connection,
+            io_loop=self.io_loop,
+        )
+        self._started = True
+
+    @tornado.gen.coroutine
+    def handle_stream(self, stream):
+        '''
+        Override this to handle the streams as they arrive
+
+        :param IOStream stream: An IOStream for processing
+
+        See http://tornado.readthedocs.org/en/latest/iostream.html#tornado.iostream.IOStream
+        for additional details.
+        '''
+        while True:
+            framed_msg_len = yield stream.read_until(' ')
+            framed_msg_raw = yield stream.read_bytes(int(framed_msg_len.strip()))
+            framed_msg = msgpack.loads(framed_msg_raw)
+            body = framed_msg['body']
+            self.io_loop.spawn_callback(self.stream_handler, body)
+
+    @tornado.gen.coroutine
+    def handle_connection(self, connection, address):
+        log.trace('IPCServer: Handling connection to address: {0}'.format(address))
+        try:
+            stream = IOStream(connection,
+                              io_loop=self.io_loop,
+                              )
+            yield self.handle_stream(stream)
+        except Exception as exc:
+            log.error('IPC streaming error: {0}'.format(exc))
+
+    def __del__(self):
+        '''
+        Routines to handle any cleanup before the instance shuts down.
+        Sockets and filehandles should be closed explicitely, to prevent
+        leaks.
+        '''
+        if hasattr(self.stream, 'close'):
+            self.stream.close()
+        if hasattr(self.sock, 'close'):
+            self.sock.close()
+
+
+class IPCClient(object):
+    '''
+    A Tornado IPC client very similar to Tornado's TCPClient class
+    but using either UNIX domain sockets or TCP sockets
+
+    This was written because Tornado does not have its own IPC
+    server/client implementation.
+
+    :param IOLoop io_loop: A Tornado ioloop to handle scheduling
+    :param str socket_path: A path on the filesystem where a socket
+                            belonging to a running IPCServer can be
+                            found.
+    '''
+    def __init__(self, io_loop=None, socket_path=None):
+        '''
+        Create a new IPC client
+
+        IPC clients cannot bind to ports, but must connect to
+        existing IPC servers. Clients can then send messages
+        to the server.
+
+        '''
+        self.io_loop = io_loop or tornado.ioloop.IOLoop.current()
+        self.socket_path = socket_path
+
+    # TODO: single connect for multiple callers (similar to auth)
+    @tornado.gen.coroutine
```

Comment:
Might want to so a singleton-esque connect-- that way if multiple people call methods that would do connect we only *actually* do one. If you want an example the auth system does this.